source/Irrlicht/FW_Mutex.h

```diff
@@ -198,7 +224,53 @@ not using add because we atomically wait for value to be 0 before swapping
 **/
 #define FW_AtomicCounterMagicBlockVal 0x1000000
 
-#if _MSC_VER && !__INTEL_COMPILER
+#if defined(FW_MUTEX_H_CXX11_IMPL)
+
+#define FW_AtomicCounter std::atomic_int
+inline void FW_AtomicCounterIncr(FW_AtomicCounter &lock)
+{
+	if (lock.fetch_add(1, std::memory_order_release)+1 > FW_AtomicCounterMagicBlockVal)
+		while (lock >= FW_AtomicCounterMagicBlockVal) ;
+}
+inline void FW_AtomicCounterDecr(FW_AtomicCounter &lock)
+{
+	lock.fetch_sub(1, std::memory_order_acquire);
+}
+inline void FW_AtomicCounterBlock(FW_AtomicCounter &lock)
+{
+	int zero = 0;
+	while (!lock.compare_exchange_weak(zero, FW_AtomicCounterMagicBlockVal, std::memory_order_acquire))
+		zero = 0;
+}
+inline void FW_AtomicCounterDecrBlock(FW_AtomicCounter &lock)
+{
+	//to make sure we get lock first, and dont keep waiting forever because every read op has priority
+	//but we check if someone else has a read or write block, in that case we're not the only ones having the lock
+	//so lock value is more than the intended FW_AtomicCounterMagicBlockVal
+	if (lock.fetch_add(FW_AtomicCounterMagicBlockVal - 1, std::memory_order_acquire) > 1)
+	{
+		//someone has a read or write block before we tried to swap access types
+		//so we release our lock completely and wait
+		lock.fetch_sub(FW_AtomicCounterMagicBlockVal, std::memory_order_acquire);
```

Comment:
more than two threads can use this atomic lock so please use std::memory_order_acq_rel

Reply:
but theoretically because it would  be sandwiched between two std::memory_order_acq_rel, we could let it be a relaxed